activesupport/lib/active_support/testing/instrumentation_assertions.rb

```diff
@@ -0,0 +1,78 @@
+      #
+      #     assert_instrumentation_event("post.submitted", title: "Cool Post") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_instrumentation_event(name, payload = nil, &block)
+        events = capture_instrumentation_events(name, &block)
+        assert_not_empty(events, "No #{name} events were found.")
+
+        return if payload.nil?
+
+        event = events.find { |event| event.payload == payload }
+        assert_not_nil(
+          event,
+          "No #{name} event with payload #{payload} was found. " \
+            "Did you mean one of these payloads: #{events.map(&:payload).join("\n")}?",
+        )
+      end
+
+      # Assert the number of events emitted with the given +name+.
+      #
+      # You can assert the number of events emitted by passing a name, count,
+      # and block. While the block is executed, the number of matching events
+      # emitted will be counted. After the block is executed, the assertion
+      # will pass if the count matches. Otherwise, it will fail.
+      #
+      #     assert_instrumentation_events_count("post.submitted", 1) do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_instrumentation_events_count(name, count, &block)
+        actual_count = capture_instrumentation_events(name, &block).count
+        assert_equal(count, actual_count, "Expected #{count} instead of #{actual_count} events for #{name}")
+      end
+
+      # Assert no events were emitted for the given +name+.
+      #
+      # You can assert no events were emitted by passing a name and block.
+      # While the block is executed, if no matching events are emitted,
+      # the assertion will pass. Otherwise, it will fail.
+      #
+      #     assert_no_instrumentation_events("post.submitted") do
+      #       post.destroy # => emits non-matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_no_instrumentation_events(name = nil, &block)
+        events = capture_instrumentation_events(name, &block)
+        error_message = name ? "Expected no events for #{name} but found #{events.size}" : "Expected no events but found #{events.size}"
+        assert_empty(events, error_message)
+      end
+
+      # Capture emitted events, optionally filtered by a +pattern+.
+      #
+      # You can capture emitted events, optionally filtered by a pattern, by passing a pattern and block.
+      #
+      #     events = capture_instrumentation_events("post.submitted") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      def capture_instrumentation_events(pattern = nil, &block)
+        events = []
+        ActiveSupport::Notifications.subscribed(->(e) { events << e }, pattern) { block.call }
```

Comment:
Does this work with `&block` as a block argument?

```suggestion
        ActiveSupport::Notifications.subscribed(->(e) { events << e }, pattern, &block)
```

Reply:
It does indeed. Nice :raised_hands: